benchmark/worker/benchmark_client.go

```diff
@@ -0,0 +1,330 @@
+	case *testpb.LoadParams_Determ:
+		return nil, grpc.Errorf(codes.Unimplemented, "unsupported load params: %v", setup.LoadParams)
+	case *testpb.LoadParams_Pareto:
+		return nil, grpc.Errorf(codes.Unimplemented, "unsupported load params: %v", setup.LoadParams)
+	default:
+		return nil, grpc.Errorf(codes.InvalidArgument, "unknown load params: %v", setup.LoadParams)
+	}
+
+	grpclog.Printf(" - rpc type: %v", setup.RpcType)
+	var rpcType string
+	switch setup.RpcType {
+	case testpb.RpcType_UNARY:
+		rpcType = "unary"
+	case testpb.RpcType_STREAMING:
+		rpcType = "streaming"
+	default:
+		return nil, grpc.Errorf(codes.InvalidArgument, "unknown rpc type: %v", setup.RpcType)
+	}
+
+	grpclog.Printf(" - histogram params: %v", setup.HistogramParams)
+	grpclog.Printf(" - server targets: %v", setup.ServerTargets)
+
+	conns := make([]*grpc.ClientConn, connCount)
+
+	for connIndex := 0; connIndex < connCount; connIndex++ {
+		conns[connIndex] = benchmark.NewClientConn(setup.ServerTargets[connIndex%len(setup.ServerTargets)], opts...)
+	}
+
+	bc := benchmarkClient{
+		histogram: stats.NewHistogram(stats.HistogramOptions{
+			NumBuckets:     int(math.Log(setup.HistogramParams.MaxPossible)/math.Log(1+setup.HistogramParams.Resolution)) + 1,
+			GrowthFactor:   setup.HistogramParams.Resolution,
+			BaseBucketSize: (1 + setup.HistogramParams.Resolution),
+			MinValue:       0,
+		}),
+		stop:          make(chan bool),
+		lastResetTime: time.Now(),
+	}
+
+	switch rpcType {
+	case "unary":
+		if dist == nil {
+			bc.doCloseLoopUnaryBenchmark(conns, rpcCountPerConn, payloadReqSize, payloadRespSize)
+		}
+		// TODO else do open loop
+	case "streaming":
+		if dist == nil {
+			bc.doCloseLoopStreamingBenchmark(conns, rpcCountPerConn, payloadReqSize, payloadRespSize, payloadType)
+		}
+		// TODO else do open loop
+	}
+
+	return &bc, nil
+}
+
+func (bc *benchmarkClient) doCloseLoopUnaryBenchmark(conns []*grpc.ClientConn, rpcCountPerConn int, reqSize int, respSize int) {
+	clients := make([]testpb.BenchmarkServiceClient, len(conns))
+	for ic, conn := range conns {
+		clients[ic] = testpb.NewBenchmarkServiceClient(conn)
+		// Do some warm up.
+		for j := 0; j < 10; j++ {
+			benchmark.DoUnaryCall(clients[ic], 1, 1)
+		}
+	}
+	for ic, conn := range conns {
+		// For each connection, create rpcCountPerConn goroutines to do rpc.
+		// Close this connection after all go routines finish.
+		var wg sync.WaitGroup
+		wg.Add(rpcCountPerConn)
+		for j := 0; j < rpcCountPerConn; j++ {
+			go func(client testpb.BenchmarkServiceClient) {
+				defer wg.Done()
+				for {
+					done := make(chan bool)
+					go func() {
+						start := time.Now()
+						if err := benchmark.DoUnaryCall(client, reqSize, respSize); err != nil {
+							done <- false
+							return
+						}
+						elapse := time.Since(start)
+						bc.mu.Lock()
+						bc.histogram.Add(int64(elapse / time.Nanosecond))
+						bc.mu.Unlock()
+						done <- true
```

Comment:
you need to read on bc.stop too so that if bc.stop is readable this goroutine will terminate.